flink-ml-lib/src/main/java/org/apache/flink/ml/classification/knn/KnnModel.java

```diff
@@ -0,0 +1,317 @@
+            if (firstElement) {
+                comparator = Comparator.comparingDouble(o -> -o.f0);
+                loadModel(getRuntimeContext().getBroadcastVariable(broadcastKey));
+                firstElement = false;
+            }
+            DenseVector vector = (DenseVector) row.getField(featureCol);
+            Tuple2<List<Double>, List<Double>> t2 = findNeighbor(vector, topN);
+            return Row.join(row, Row.of(getResult(t2)));
+        }
+
+        /**
+         * Finds the nearest topN neighbors from whole nodes.
+         *
+         * @param input Input vector.
+         * @param topN Top N.
+         * @return Neighbors.
+         */
+        private Tuple2<List<Double>, List<Double>> findNeighbor(DenseVector input, Integer topN) {
+            PriorityQueue<Tuple2<Double, Double>> priorityQueue = new PriorityQueue<>(comparator);
+            search(input, topN, priorityQueue);
+            List<Double> items = new ArrayList<>();
+            List<Double> metrics = new ArrayList<>();
+            while (!priorityQueue.isEmpty()) {
+                Tuple2<Double, Double> result = priorityQueue.poll();
+                items.add(result.f1);
+                metrics.add(result.f0);
+            }
+            Collections.reverse(items);
+            Collections.reverse(metrics);
+            priorityQueue.clear();
+            return Tuple2.of(items, metrics);
+        }
+
+        /**
+         * @param input Input vector.
+         * @param topN Top N.
+         * @param priorityQueue Priority queue.
+         */
+        private void search(
+                DenseVector input, int topN, PriorityQueue<Tuple2<Double, Double>> priorityQueue) {
+            double d = 0.0;
+            for (int i = 0; i < input.size(); ++i) {
+                d += input.values[i] * input.values[i];
+            }
+            Tuple2<DenseVector, Double> sample = Tuple2.of(input, d);
+            Tuple2<Double, Double> head = null;
+
+            List<Tuple2<Double, Double>> values = computeDistance(sample);
+            for (Tuple2<Double, Double> currentValue : values) {
+                head = updateQueue(priorityQueue, topN, currentValue, head);
+            }
+        }
+
+        /**
+         * Updates queue.
+         *
+         * @param pq Queue.
+         * @param topN Top N.
+         * @param newValue New value.
+         * @param head Head value.
+         * @return Head value.
+         */
+        private <T> Tuple2<Double, T> updateQueue(
+                PriorityQueue<Tuple2<Double, T>> pq,
+                int topN,
+                Tuple2<Double, T> newValue,
+                Tuple2<Double, T> head) {
+            if (pq.size() < topN) {
+                pq.add(Tuple2.of(newValue.f0, newValue.f1));
+                head = pq.peek();
+            } else {
+                if (pq.comparator().compare(head, newValue) < 0) {
+                    Tuple2<Double, T> peek = pq.poll();
+                    assert peek != null;
```

Comment:
Could we avoid using `assert` in production code? `assert` is not enabled by default in production code.

Reply:
OK